include/openmc/random_ray/parallel_map.h

```diff
@@ -0,0 +1,201 @@
+    {
+      ++inner_it_;
+      advance_to_valid();
+      return *this;
+    }
+
+    iterator operator++(int)
+    {
+      iterator tmp = *this;
+      ++(*this);
+      return tmp;
+    }
+
+    bool operator==(const iterator& other) const
+    {
+      // Two iterators are equal if they refer to the same bucket vector and are
+      // both at end, or if they have the same bucket index and inner iterator.
+      return buckets_ == other.buckets_ &&
+             bucket_index_ == other.bucket_index_ &&
+             (bucket_index_ == buckets_->size() ||
+               inner_it_ == other.inner_it_);
+    }
+
+    bool operator!=(const iterator& other) const { return !(*this == other); }
+
+  private:
+    // Helper function: if we are at the end of the current bucket, advance to
+    // the next non-empty bucket.
+    void advance_to_valid()
+    {
+      while (bucket_index_ < buckets_->size() &&
+             inner_it_ == (*buckets_)[bucket_index_].map_.end()) {
+        ++bucket_index_;
+        if (bucket_index_ < buckets_->size())
+          inner_it_ = (*buckets_)[bucket_index_].map_.begin();
+      }
+    }
+
+    std::vector<Bucket>* buckets_;
+    std::size_t bucket_index_;
+    typename std::unordered_map<KeyType, std::unique_ptr<ValueType>,
+      HashFunctor>::iterator inner_it_;
+  };
+
+public:
+  //----------------------------------------------------------------------------
+  // Constructor
+  ParallelMap(int n_buckets = 1000) : buckets_(n_buckets) {}
+
+  //----------------------------------------------------------------------------
+  // Public Methods
+  void lock(const KeyType& key)
+  {
+    Bucket& bucket = get_bucket(key);
+    bucket.lock_.lock();
+  }
+
+  void unlock(const KeyType& key)
+  {
+    Bucket& bucket = get_bucket(key);
+    bucket.lock_.unlock();
+  }
+
+  void clear()
+  {
+    for (auto& bucket : buckets_) {
+      bucket.map_.clear();
+    }
+  }
+
+  bool contains(const KeyType& key)
+  {
+    Bucket& bucket = get_bucket(key);
+    // C++20
+    // return bucket.map_.contains(key);
+    return bucket.map_.find(key) != bucket.map_.end();
+  }
+
+  ValueType& operator[](const KeyType& key)
+  {
+    Bucket& bucket = get_bucket(key);
+    return *bucket.map_[key].get();
+  }
+
+  ValueType* emplace(KeyType& key, const ValueType& value)
+  {
+    Bucket& bucket = get_bucket(key);
+    // Attempt to emplace the new element into the unordered_map within the
+    auto result = bucket.map_.emplace(key, std::make_unique<ValueType>(value));
+    auto it = result.first;
+    return it->second.get();
+  }
+
+  // Copies everything into a vector, clears all buckets
+  int64_t move_contents_into_vector(vector<ValueType>& v)
+  {
+    int64_t n_new = 0;
+    for (auto& bucket : buckets_) {
+      for (auto& pair : bucket.map_) {
+        v.push_back(*pair.second.get());
+        n_new++;
+      }
+      bucket.map_.clear();
+    }
+    return n_new;
+  }
+
+  // Return iterator to first element.
+  iterator begin()
```

Comment:
I'm assuming iteration is always done at a point where the map is not being modified?

Reply:
That's right. I've added in some more detail to the longer comment block before the class that describes the point of the serial iterators. Yes, the iterators are just an abstract way of dumping everything out from the class (in serial), rather than having to make the user create their own loop to navigate all the bins.